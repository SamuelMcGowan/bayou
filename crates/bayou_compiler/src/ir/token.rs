use bayou_diagnostic::span::Span;

use super::NodeCopy;
use crate::ir::InternedStr;

#[derive(NodeCopy!)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum TokenKind {
    Keyword(Keyword),
    Identifier(InternedStr),
    Integer(i64),

    LBrace,
    RBrace,
    LParen,
    RParen,
    Semicolon,
    Bang,
    Arrow,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Assign,
    // TODO: uncomment when assignment is added
    // AddEq,
    // SubEq,
    // MulEq,
    // DivEq,
    // ModEq,

    // TODO: uncomment when boolean operations are added
    // EqEq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub enum Keyword {
    Func,
    Return,

    Let,

    Void,
    I64,
}

impl TokenKind {
    pub fn token_name(&self) -> &'static str {
        match self {
            TokenKind::Keyword(kw) => match kw {
                Keyword::Func => "keyword `func`",
                Keyword::Return => "keyword `return`",
                Keyword::Let => "keyword `let`",
                Keyword::I64 => "keyword `i64`",
                Keyword::Void => "keyword `void`",
            },
            TokenKind::Identifier(_) => "identifier",
            TokenKind::Integer(_) => "integer",
            TokenKind::LBrace => "`{`",
            TokenKind::RBrace => "`}`",
            TokenKind::LParen => "`(`",
            TokenKind::RParen => "`)`",
            TokenKind::Semicolon => "`;`",
            TokenKind::Bang => "`!`",
            TokenKind::Arrow => "`->`",
            TokenKind::Add => "`+`",
            TokenKind::Sub => "`-`",
            TokenKind::Mul => "`*`",
            TokenKind::Div => "`/`",
            TokenKind::Mod => "`%`",
            TokenKind::Assign => "`=`",
            TokenKind::BitwiseAnd => "`&`",
            TokenKind::BitwiseOr => "`|`",
            TokenKind::BitwiseXor => "`^`",
            TokenKind::BitwiseInvert => "`~`",
        }
    }
}
