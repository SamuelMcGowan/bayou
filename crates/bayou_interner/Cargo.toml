[package]
name = "bayou_interner"
version = "0.1.0"
edition = "2021"

[dependencies]
ahash = "0.8.11"
bumpalo = "3.16.0"
hashbrown = "0.14.5"

[dependencies.serde]
workspace = true
optional = true
features = ["derive"]
