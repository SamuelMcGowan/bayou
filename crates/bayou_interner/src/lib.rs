//! A string interner.
//!
//! A string interner is a data structure commonly used in compilers and other contexts that need to
//! cheaply store and compare many often identical strings. "Interning" a string returns a pointer (or in
//! this implementation, an ID) that is cheap to copy and to perform string equality checks on. This is
//! achieved by deduplicating strings using an internal hash table.
//!
//! This string interner also stores all strings in a single bump-allocated buffer, courtesy of [`bumpalo`],
//! avoiding excessive allocation.
//!
//! I decided to represent interned strings with a 32-bit ID instead of a reference to avoid introducing lifetimes.
//! This does mean that accessing the underlying string requires calling a method on the interner, but this is a
//! single array lookup.
//!
//! # Example
//! ```rust
//! use bayou_interner::Interner;
//!
//! # fn main() {
//! let interner = Interner::new();
//!
//! let hello = interner.intern("hello");
//! let hello2 = interner.intern("hello");
//! let world = interner.intern("world");
//!
//! // Interned strings can be compared cheaply.
//! assert_ne!(hello, hello2);
//! assert_ne!(hello, world);
//!
//! // Getting the associated string for an interned string.
//! assert_eq!(interner.get_str(hello), Some("hello"));
//! # }
//! ```

mod arena;

use std::cell::RefCell;
use std::num::NonZeroU32;
use std::ops::Index;

use ahash::RandomState;
use arena::InternerArena;
use hashbrown::hash_table::Entry;
use hashbrown::HashTable;

/// An ID for an interned string. Cheap to copy, and to perform string equality checks on, as
/// internally it is simply a [`NonZeroU32`] ID. It can also be stored inside an [`Option`] for free
/// due to niche optimisation.
///
/// In order to get the associated string, the interned string must be looked up
/// in the interner it was created with.
///
/// Note that performing an equality check on interned strings from different
/// interners will give a nonsensical result.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Istr(NonZeroU32);

impl Istr {
    #[inline]
    fn from_index(index: usize) -> Option<Self> {
        Some(Self(NonZeroU32::new(
            u32::try_from(index).ok()?.wrapping_add(1),
        )?))
    }

    #[inline]
    fn to_index(self) -> usize {
        self.0.get() as usize - 1
    }
}

#[derive(Clone, Copy)]
struct Metadata {
    interned: Istr,
    hash: u64,
}

#[derive(Default)]
struct Lookup {
    random_state: RandomState,
    table: HashTable<Metadata>,
}

/// Storage for interned strings.
#[derive(Default)]
pub struct Interner {
    lookup: RefCell<Lookup>,
    arena: InternerArena,
}

impl Interner {
    /// Create a new interner.
    #[inline]
    pub fn new() -> Self {
        Self::default()
    }

    /// Intern a string, returning an interned string that it is cheap to copy and
    /// perform equality checks on. Strings are only stored in the interner once, no
    /// matter how many times they are interned.
    ///
    ///
    /// ```rust
    /// # use bayou_interner::Interner;
    /// #
    /// # fn main() {
    /// let interner = Interner::new();
    ///
    /// let hello = interner.intern("hello");
    /// let hello2 = interner.intern("hello");
    /// let world = interner.intern("world");
    ///
    /// assert_eq!(hello, hello2);
    /// assert_ne!(hello, world);
    /// # }
    /// ```
    ///
    /// # Panics
    /// Panics if there are no more available IDs. An interner can store up
    /// to `u32::MAX - 1` strings before panicking.
    #[inline]
    pub fn intern(&self, key: &str) -> Istr {
        self.try_intern(key).expect("Too many interned strings")
    }

    /// Like [`Interner::intern`], but non-panicking in the case that there are no
    /// more available IDs.
    pub fn try_intern(&self, key: &str) -> Option<Istr> {
        let mut lookup = self.lookup.borrow_mut();

        let hash = lookup.random_state.hash_one(key);

        let entry = lookup.table.entry(
            hash,
            |metadata| self.arena.get(metadata.interned.to_index()) == Some(key),
            |metadata| metadata.hash,
        );

        let interned = match entry {
            Entry::Occupied(entry) => entry.get().interned,
            Entry::Vacant(entry) => {
                let index = self.arena.push_str(key);
                let interned = Istr::from_index(index)?;

                entry.insert(Metadata { interned, hash });

                interned
            }
        };

        Some(interned)
    }

    /// Get an interned string if this string is interned, otherwise return `None`.
    ///
    /// ```rust
    /// # use bayou_interner::{Interner, Istr};
    /// #
    /// # fn main() {
    /// let interner = Interner::new();
    ///
    /// let hello: Istr = interner.intern("hello");
    ///
    /// assert_eq!(interner.get_interned("hello"), Some(hello));
    /// assert_eq!(interner.get_interned("world"), None);
    /// # }
    /// ```
    pub fn get_interned(&self, key: &str) -> Option<Istr> {
        let lookup = self.lookup.borrow();

        let hash = lookup.random_state.hash_one(key);

        lookup
            .table
            .find(hash, |metadata| {
                self.arena.get(metadata.interned.to_index()) == Some(key)
            })
            .map(|metadata| metadata.interned)
    }

    /// Look up an interned string to get the associated string.
    ///
    /// Note that if the interned string was created by another interner
    /// this method will return an arbitrary string or `None`.
    ///
    /// If you know that an interned string was created by this interner, you can index into
    /// the interner instead. This is the same as calling this method, but panics if the
    /// interned string is not found in this interner instead of returning `None`.
    ///
    /// ```rust
    /// # use bayou_interner::Interner;
    /// #
    /// # fn main() {
    /// let interner = Interner::new();
    /// let interned = interner.intern("hello");
    ///
    /// assert_eq!(interner.get_str(interned), Some("hello"));
    /// # }
    /// ```
    #[inline]
    pub fn get_str(&self, interned: Istr) -> Option<&str> {
        self.arena.get(interned.to_index())
    }
}

impl Index<Istr> for Interner {
    type Output = str;

    #[inline]
    fn index(&self, interned: Istr) -> &Self::Output {
        self.get_str(interned).expect("String not in interner")
    }
}

#[test]
fn test_interner() {
    let interner = Interner::new();

    for n in 0..100 {
        let s = n.to_string();

        let a = interner.intern(&s);
        let b = interner.intern(&s);

        assert_eq!(a, b);
        assert_eq!(interner.get_str(a), Some(s.as_str()));
    }
}
