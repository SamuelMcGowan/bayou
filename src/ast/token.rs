use crate::session::InternedStr;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Token {
    Keyword(Keyword),
    Identifier(InternedStr),
    Integer(u64),

    LBrace,
    RBrace,
    LParen,
    RParen,
    Semicolon,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Keyword {
    Int,
    Return,
}
