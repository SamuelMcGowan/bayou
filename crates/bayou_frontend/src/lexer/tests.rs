use bayou_interner::Interner;
use bayou_utils::assert_yaml_snapshot_with_source;

use super::{Lexer, LexerError};
use crate::token::Token;

fn lex(source: &str) -> (Vec<Token>, Vec<LexerError>) {
    let mut interner = Interner::new();

    let lexer = Lexer::new(source, &mut interner);
    let (tokens, lexer_errors) = lexer.lex();

    (tokens.collect(), lexer_errors)
}

macro_rules! assert_lex {
    ($source:expr) => {{
        let source = $source;
        assert_yaml_snapshot_with_source!(source => lex(source));
    }};
}

#[test]
fn integer() {
    assert_lex!("100");
}

#[test]
fn integer_overflow() {
    assert_lex!("100000000000000000000");
}

#[test]
fn newlines() {
    assert_lex!("\nfunc\nmain\n(\n)\n->\ni64\n{\nreturn\n0\n;\n}");
}

#[test]
fn no_newlines() {
    assert_lex!("func main()->i64{return 0;}");
}

#[test]
fn spaces() {
    assert_lex!("  func  main  (  )  ->  i64  {  return  0  ;  }");
}
