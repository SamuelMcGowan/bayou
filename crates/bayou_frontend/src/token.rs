use bayou_diagnostic::span::Span;
use bayou_interner::Istr;

use crate::NodeCopy;

#[derive(NodeCopy!)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum TokenKind {
    Keyword(Keyword),
    Identifier(Istr),
    Integer(i64),
    Bool(bool),

    LBrace,
    RBrace,
    LParen,
    RParen,

    Dot,
    Colon,
    Comma,
    Semicolon,
    Bang,
    Arrow,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Assign,
    // TODO: uncomment when assignment is added
    // AddEq,
    // SubEq,
    // MulEq,
    // DivEq,
    // ModEq,

    // TODO: uncomment when boolean operations are added
    // EqEq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub enum Keyword {
    Func,
    Return,

    Let,
    If,
    Then,
    Else,

    I64,
    Bool,
    Void,
}

impl TokenKind {
    pub fn token_name(&self) -> &'static str {
        match self {
            TokenKind::Keyword(kw) => match kw {
                Keyword::Func => "keyword `func`",
                Keyword::Return => "keyword `return`",
                Keyword::Let => "keyword `let`",
                Keyword::If => "keyword `if`",
                Keyword::Then => "keyword `then`",
                Keyword::Else => "keyword `else`",
                Keyword::I64 => "keyword `i64`",
                Keyword::Bool => "keyword `bool`",
                Keyword::Void => "keyword `void`",
            },
            TokenKind::Identifier(_) => "identifier",
            TokenKind::Integer(_) => "integer",
            TokenKind::Bool(_) => "boolean",
            TokenKind::LBrace => "`{`",
            TokenKind::RBrace => "`}`",
            TokenKind::LParen => "`(`",
            TokenKind::RParen => "`)`",
            TokenKind::Dot => "`.`",
            TokenKind::Colon => "`:`",
            TokenKind::Comma => "`,`",
            TokenKind::Semicolon => "`;`",
            TokenKind::Bang => "`!`",
            TokenKind::Arrow => "`->`",
            TokenKind::Add => "`+`",
            TokenKind::Sub => "`-`",
            TokenKind::Mul => "`*`",
            TokenKind::Div => "`/`",
            TokenKind::Mod => "`%`",
            TokenKind::Assign => "`=`",
            TokenKind::BitwiseAnd => "`&`",
            TokenKind::BitwiseOr => "`|`",
            TokenKind::BitwiseXor => "`^`",
            TokenKind::BitwiseInvert => "`~`",
        }
    }
}
