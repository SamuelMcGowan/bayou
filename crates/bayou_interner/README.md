A string interner.

A string interner is a data structure commonly used in compilers and other contexts that need to
cheaply store and compare many often identical strings. "Interning" a string returns a pointer (or in
this implementation, an ID) that is cheap to copy and to perform string equality checks on. This is
achieved by deduplicating strings using an internal hash table.

This string interner also stores all strings in a single bump-allocated buffer, courtesy of
[bumpalo](https://crates.io/crates/bumpalo), avoiding excessive allocation.

I decided to represent interned strings with a 32-bit ID instead of a reference to avoid introducing lifetimes.
This does mean that accessing the underlying string requires calling a method on the interner, but this is a
single array lookup.

# Example
```rust
use bayou_interner::Interner;

let interner = Interner::new();

let hello = interner.intern("hello");
let hello2 = interner.intern("hello");
let world = interner.intern("world");

// Interned strings can be compared cheaply.
assert_ne!(hello, hello2);
assert_ne!(hello, world);

// Getting the associated string for an interned string.
assert_eq!(interner.get_str(hello), Some("hello"));
```