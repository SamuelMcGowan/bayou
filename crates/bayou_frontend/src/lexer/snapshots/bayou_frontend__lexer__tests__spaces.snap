---
source: crates/bayou_frontend/src/lexer/tests.rs
info: "  func  main  (  )  ->  i64  {  return  0  ;  }"
---
- - kind:
      Keyword: Func
    span:
      start: 2
      end: 6
  - kind:
      Identifier: 1
    span:
      start: 8
      end: 12
  - kind: LParen
    span:
      start: 14
      end: 15
  - kind: RParen
    span:
      start: 17
      end: 18
  - kind: Arrow
    span:
      start: 20
      end: 22
  - kind:
      Keyword: I64
    span:
      start: 24
      end: 27
  - kind: LBrace
    span:
      start: 29
      end: 30
  - kind:
      Keyword: Return
    span:
      start: 32
      end: 38
  - kind:
      Integer: 0
    span:
      start: 40
      end: 41
  - kind: Semicolon
    span:
      start: 43
      end: 44
  - kind: RBrace
    span:
      start: 46
      end: 47
- []
